File: backend/src/main/kotlin/com/otoki/internal/entity/Product.kt
package com.otoki.internal.entity

import jakarta.persistence.*
import java.time.LocalDateTime

/**
 * 제품 Entity
 * Orora 영업 시스템에서 동기화되는 제품 마스터 데이터
 */
@Entity
@Table(
    name = "products",
    indexes = [
        Index(name = "idx_products_product_id", columnList = "product_id", unique = true),
        Index(name = "idx_products_product_code", columnList = "product_code"),
        Index(name = "idx_products_barcode", columnList = "barcode"),
        Index(name = "idx_products_product_name", columnList = "product_name")
    ]
)
class Product(

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0,

    @Column(name = "product_id", nullable = false, unique = true, length = 20)
    val productId: String,

    @Column(name = "product_name", nullable = false, length = 200)
    val productName: String,

    @Column(name = "product_code", nullable = false, length = 20)
    val productCode: String,

    @Column(name = "barcode", nullable = false, length = 20)
    val barcode: String,

    @Column(name = "storage_type", nullable = false, length = 10)
    val storageType: String,

    @Column(name = "shelf_life", length = 20)
    val shelfLife: String? = null,

    @Column(name = "category_mid", length = 50)
    val categoryMid: String? = null,

    @Column(name = "category_sub", length = 50)
    val categorySub: String? = null,

    @Column(name = "pieces_per_box", nullable = false)
    val piecesPerBox: Int = 1,

    @Column(name = "min_order_unit", nullable = false)
    val minOrderUnit: Int = 1,

    @Column(name = "supply_quantity", nullable = false)
    val supplyQuantity: Int = 0,

    @Column(name = "dc_quantity", nullable = false)
    val dcQuantity: Int = 0,

    @Column(name = "unit_price", nullable = false)
    val unitPrice: Long = 0,

    @Column(name = "created_at", nullable = false, updatable = false)
    val createdAt: LocalDateTime = LocalDateTime.now(),

    @Column(name = "updated_at", nullable = false)
    var updatedAt: LocalDateTime = LocalDateTime.now()
) {

    @PreUpdate
    fun onPreUpdate() {
        this.updatedAt = LocalDateTime.now()
    }
}

File: mobile/lib/domain/entities/product.dart
/// 제품 엔티티
///
/// 제품검색 결과에 표시되는 제품 정보를 담는 도메인 엔티티입니다.
class Product {
  /// 제품 ID (고유 식별자)
  final String productId;

  /// 제품명
  final String productName;

  /// 제품코드
  final String productCode;

  /// 바코드
  final String barcode;

  /// 보관 조건 (냉장/냉동/상온)
  final String storageType;

  /// 유통기한 기간
  final String shelfLife;

  /// 중분류 카테고리
  final String? categoryMid;

  /// 소분류 카테고리
  final String? categorySub;

  const Product({
    required this.productId,
    required this.productName,
    required this.productCode,
    required this.barcode,
    required this.storageType,
    required this.shelfLife,
    this.categoryMid,
    this.categorySub,
  });

  Product copyWith({
    String? productId,
    String? productName,
    String? productCode,
    String? barcode,
    String? storageType,
    String? shelfLife,
    String? categoryMid,
    String? categorySub,
  }) {
    return Product(
      productId: productId ?? this.productId,
      productName: productName ?? this.productName,
      productCode: productCode ?? this.productCode,
      barcode: barcode ?? this.barcode,
      storageType: storageType ?? this.storageType,
      shelfLife: shelfLife ?? this.shelfLife,
      categoryMid: categoryMid ?? this.categoryMid,
      categorySub: categorySub ?? this.categorySub,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'productId': productId,
      'productName': productName,
      'productCode': productCode,
      'barcode': barcode,
      'storageType': storageType,
      'shelfLife': shelfLife,
      'categoryMid': categoryMid,
      'categorySub': categorySub,
    };
  }

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      productId: json['productId'] as String,
      productName: json['productName'] as String,
      productCode: json['productCode'] as String,
      barcode: json['barcode'] as String,
      storageType: json['storageType'] as String,
      shelfLife: json['shelfLife'] as String,
      categoryMid: json['categoryMid'] as String?,
      categorySub: json['categorySub'] as String?,
    );
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is Product &&
        other.productId == productId &&
        other.productName == productName &&
        other.productCode == productCode &&
        other.barcode == barcode &&
        other.storageType == storageType &&
        other.shelfLife == shelfLife &&
        other.categoryMid == categoryMid &&
        other.categorySub == categorySub;
  }

  @override
  int get hashCode {
    return Object.hash(
      productId,
      productName,
      productCode,
      barcode,
      storageType,
      shelfLife,
      categoryMid,
      categorySub,
    );
  }

  @override
  String toString() {
    return 'Product(productId: $productId, productName: $productName, '
        'productCode: $productCode, barcode: $barcode, '
        'storageType: $storageType, shelfLife: $shelfLife, '
        'categoryMid: $categoryMid, categorySub: $categorySub)';
  }
}
